Partitioned topic and topic negotiation

cammellos opened the pull request:
Adds topic negotiation and partitioned topic for group-chats and pairing.

All the code has been implemented in statusgo: https://github.com/status-im/status-go/pull/1466

Basically all the whisper filter management is done at that level.

## Technical description

* On startup we load all chats and send a list of them to status go:

For a public chat: `{:chatId "status"}`, we create a single filter, based on the name of the chat.

For each contact added by us, each user in a group chat and each one to one chat open, we send:
`{:chatId "0x", :oneToOne true}`. This will create a chats, to listen to their contact code.

Any previously negotiated topic is also returned.

* Once loaded, we create our filters, and upsert the mailserver topics, both of which are solely based on the filters loaded.

In order to remove a chat, we delete/stopwatching first the the filter in status-react and then ask status-go to remove the filter. For a public chat we always remove, for a one-to-one we remove only if the user is not in our contacts, or in a group chat or we have a chat open. Negotiated topics are never removed, as otherwise the other user won't be able to contact us anymore.

On stopping whisper we don't have to ask status-go to remove filters as they are removed automatically.

Some more logic can be pushed in status-go, but that will be in subsequent PRs.

## Testing

This is a fairly large change and affects quite a large part of the codebase.
Specifically:

1) Group chats
2) Pairing
3) Retrieving history from mailserver
4) Creating/restoring accounts
5) Gaps
6) Compatibility with previous versions
7) Deleting/Creating one-to-one/group/public chats
8)Performance just after login (between login and plus icon is visible), it should be slightly faster if anything but roughly there should be no difference

Everything should work just in the same way.

To test the actual feature you will have check `geth.log`, in Debug level, for lines.

`Sending on negotiated topic`

`Sending on partitioned topic`

`Sending on old discovery topic`

And also you need to make sure the public key printed after is the one for the right contact (we also send pairing messages, to your own public key).

## How it works

It's all based on device-to-device communications, so make sure you pair your devices before testing (might work even if you don't pair them), and that you don't have more than 3 devices on each side.

The idea is that given a group of devices A1,A2,A3 you will send on:

`old discovery topic` that happens if 1) if you don't have information about their devices (no communication happened before) or  2) one of those devices is running an older version.

`partitioned topic` if all 3 devices are running this version, but you have not send a message back from *each one* of those devices.

`negotiated topic` if all 3 devices are running this version and you have sent a message back from each one of those devices.

## Examples 

#### 1 device on each side

A1 running this version, B1 running this version:

Send messages from A1, (don't send any message from B1) you should see either:  `Sending on old discovery topic` or `Sending on partitioned topic`. The more you wait with both devices online, the more likely you will start seeing `Sending on partitioned topic`. (This is due to the fact that devices advertise themselves periodically, so it might take some time to detect each other).

Once you reply from B1 you should see `Sending on negotiated topic` on both devices when sending a message.

A1 running this version, B1 running an older version

You should always see `Sending on old discovery topic` when sending messages from A1 (B1 will not have those logs as running an older version).

### 1 device on one side, 2 on the other

#### A1 running this version, B1 , B2 running this version.

Pair B1 & B2, sending from A1 should be either `Sending on old discovery topic` or `Sending on partitioned topic`. 
Send a message from B1, you should see `Send on negotiated topic`.
Send a message from A1 you should see `Send on partitioned topic`.
Send a message from B2 you should see `Send on negotiated topic`.
Send a message from A1 you should see `Send on negotiated topic`.

#### A1 running this version, B1  running this version B2 running an older version.

Pair B1 & B2, sending from A1 should be either `Sending on old discovery topic` or `Sending on partitioned topic`. 
Send a message from B1, you should see `Send on negotiated topic`.
Send a message from A1 you should see `Send on old discovery topic`.
Send a message from B2 (no logs as an older version).
Send a message from A1 you should see `Send on old discovery topic`.

If a device is running this version, and sending a message to a user that has at least one device running an older version, you should always see: `Sending on old discovery topic`.

Basically if all the devices you are sending the message to are running this version

apologies if it's not clear, feel free to ask any question.

To test using normal one to one messages (easier) you can enable `device-to-device` in the settings.

status: ready

cammellos committed:
Adds topic negotiation and partitioned topic

All the code has been implemented in statusgo: status-im/status-go#1466

Basically all the whisper filter management is done at that level.

Technical description
On startup we load all chats and send a list of them to status go:
For a public chat: {:chatId "status"}, we create a single filter, based on the name of the chat.

For each contact added by us, each user in a group chat and each one to one chat open, we send:
{:chatId "0x", :oneToOne true}. This will create a chats, to listen to their contact code.

Any previously negotiated topic is also returned.

Once loaded, we create our filters, and upsert the mailserver topics, both of which are solely based on the filters loaded.
In order to remove a chat, we delete/stopwatching first the the filter in status-react and then ask status-go to remove the filter. For a public chat we always remove, for a one-to-one we remove only if the user is not in our contacts, or in a group chat or we have a chat open. Negotiated topics are never removed, as otherwise the other user won't be able to contact us anymore.

On stopping whisper we don't have to ask status-go to remove filters as they are removed automatically.

Some more logic can be pushed in status-go, but that will be in subsequent PRs.

Signed-off-by: Andrea Maria Piana <andrea.maria.piana@gmail.com>

status-im-auto commented:
### Jenkins Builds
| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :x: | f0a2999a | [#1](https://ci.status.im/job/status-react/job/prs/job/windows/job/PR-8363/1/) | 2019-06-05 11:48:44 | ~4 min | `windows` | [:page_facing_up: log](https://ci.status.im/job/status-react/job/prs/job/windows/job/PR-8363/1/consoleText) |

yenda commented on modules/react-native-status/android/src/main/java/im/status/ethereum/module/StatusModule.java:
```diff
@@ -1133,4 +1133,64 @@ public void run() {
+  }
+
+  @ReactMethod
+  public void removeFilter(final String chat, final Callback callback) {
```
why did you chose to do this in native rather than using rpc?

cammellos replied:
yes, I went initially for native, I got my signals crossed, I thought that was what we wanted, but it's on RPC already, I will remove this

yenda commented on src/status_im/chat/models.cljs:
```diff
@@ -169,22 +169,16 @@
                                   (log/error "can't remove a chat:" error))}]}
               (navigation/navigate-to-cofx :home {}))
     (fx/merge cofx
-              ;; TODO: There's a race condition here, as the removal of the filter (async)
```
nice!

yenda commented on src/status_im/contact_code/core.cljs:
```diff
@@ -17,9 +17,9 @@
 
 (fx/defn listen [cofx chat-id]
-  (transport.public-chat/join-public-chat
+  (transport.public-chat/join-one-to-one-chat
```
that's odd why is join one-to-one in public-chat namespace

cammellos replied:
oh it's all still WIP, I just did not want to modify `join-public-chat` for now

yenda commented on src/status_im/mailserver/core.cljs:
```diff
@@ -34,7 +34,8 @@
 
 
-(def one-day (* 24 3600))
+(def one-hour 3600)
```
seriously :D

cammellos replied:
testing :)

yenda commented on src/status_im/native_module/impl/module.cljs:
```diff
@@ -152,6 +152,30 @@
   (when (status)
     (.disableInstallation (status) installation-id callback)))
 
+(defn load-filters [chats callback]
```
ok here it is rpc calls so the native code above is going to be removed?

cammellos replied:
correct

yenda commented on src/status_im/transport/filters.cljs:
```diff
@@ -22,14 +22,14 @@
     (keep
      (fn [{:keys [options callback chat-id]}]
-       (when-let [filter (.newMessageFilter
+       (when-let [filter (.newRawMessageFilter
```
seems like the last use of web3 could you use json rpc here so we can finaly get rid of it?

cammellos replied:
this one I need for the polling (I just need the `Filter` js object), it actually does not make any request through RPC, but we can get rid of it all together once we use signals for messages, once we get closer to completion of the PR we can revisit and we can decide whether to include it in this one or a separate, depending on how much stuff there is to test

jakubgs reviewed (approved):
I don't see anything wrong with the dependency changes, but this bot is really starting to get on my nerves with how much it spams PRs, @PombeirP any thoughts on maybe quieting it down?

status-im-auto commented:
#### :heavy_check_mark: [status-react/prs/ios/PR-8363#12](https://ci.status.im/job/status-react/job/prs/job/ios/job/PR-8363/12/) :small_blue_diamond: ~11 min :small_blue_diamond: f78dde55 :small_blue_diamond: [:package: ios package](https://status-im-prs.ams3.digitaloceanspaces.com/StatusIm-190612-092354-f78dde-pr.ipa)

asemiankevich commented:
@cammellos seems all builds are failed after last force push on tests step, please take a look

```
-im.wallet.db/too-precise-amount? is not public at line 17 /home/jenkins/workspace/status-react_prs_android_PR-8363/test/cljs/status_im/test/ui/screens/wallet/db.cljs
16:37:10  WARNING: var: status-im.wallet.db/too-precise-amount? is not public at line 18 /home/jenkins/workspace/status-react_prs_android_PR-8363/test/cljs/status_im/test/ui/screens/wallet/db.cljs
16:37:10  WARNING: var: status-im.wallet.db/too-precise-amount? is not public at line 19 /home/jenkins/workspace/status-react_prs_android_PR-8363/test/cljs/status_im/test/ui/screens/wallet/db.cljs
16:37:10  WARNING: var: status-im.wallet.db/too-precise-amount? is not public at line 20 /home/jenkins/workspace/status-react_prs_android_PR-8363/test/cljs/status_im/test/ui/screens/wallet/db.cljs
16:37:11  WARNING: var: status-im.transport.filters.core/chats->filter-requests is not public at line 62 /home/jenkins/workspace/status-react_prs_android_PR-8363/test/cljs/status_im/test/transport/filters/core.cljs
16:37:11  WARNING: var: status-im.transport.filters.core/chats->filter-requests is not public at line 68 /home/jenkins/workspace/status-react_prs_android_PR-8363/test/cljs/status_im/test/transport/filters/core.cljs
16:37:11  WARNING: var: status-im.transport.filters.core/chats->filter-requests is not public at line 79 /home/jenkins/workspace/status-react_prs_android_PR-8363/test/cljs/status_im/test/transport/filters/core.cljs
16:37:11  WARNING: var: status-im.transport.filters.core/contacts->filter-requests is not public at line 89 /home/jenkins/workspace/status-react_prs_android_PR-8363/test/cljs/status_im/test/transport/filters/core.cljs
16:37:11  WARNING: var: status-im.subs/get-balance-total-value is not public at line 7 /home/jenkins/workspace/status-react_prs_android_PR-8363/test/cljs/status_im/test/wallet/subs.cljs
16:37:11  WARNING: var: status-im.extensions.capacities.ethereum/generate-topic is not public at line 11 test/cljs/status_im/test/extensions/ethereum.cljs
16:37:11  WARNING: var: status-im.extensions.capacities.ethereum/generate-topic is not public at line 18 test/cljs/status_im/test/extensions/ethereum.cljs
16:37:11  WARNING: var: status-im.extensions.capacities.ethereum/generate-topic is not public at line 24 test/cljs/status_im/test/extensions/ethereum.cljs
16:37:11  WARNING: var: status-im.extensions.capacities.ethereum/generate-topic is not public at line 30 test/cljs/status_im/test/extensions/ethereum.cljs
16:37:11  WARNING: var: status-im.extensions.capacities.ethereum/generate-topic is not public at line 37 test/cljs/status_im/test/extensions/ethereum.cljs
16:37:11  WARNING: var: status-im.extensions.capacities.ethereum/ensure-hex-bn is not public at line 41 test/cljs/status_im/test/extensions/ethereum.cljs
16:37:11  WARNING: var: status-im.extensions.capacities.ethereum/ensure-hex-bn is not public at line 44 test/cljs/status_im/test/extensions/ethereum.cljs
16:37:11  WARNING: var: status-im.extensions.capacities.ethereum/ensure-hex-bn is not public at line 47 test/cljs/status_im/test/extensions/ethereum.cljs
16:37:11  WARNING: var: status-im.extensions.capacities.ethereum/ensure-hex-bn is not public at line 50 test/cljs/status_im/test/extensions/ethereum.cljs
16:37:14  WARNING: Use of undeclared Var status-im.react-native.js-dependencies/react-native-shake at line 48 src/status_im/android/core.cljs
16:37:14  clojure.lang.ExceptionInfo: failed compiling file:src/status_im/contact_code/core.cljs {:file #object[java.io.File 0x41cc5e46 "src/status_im/contact_code/core.cljs"], :clojure.error/phase :compilation}
16:37:14  	at cljs.compiler$compile_file$fn__3901.invoke(compiler.cljc:1706)
16:37:14  	at cljs.compiler$compile_file.invokeStatic(compiler.cljc:1666)
16:37:14  	at cljs.compiler$compile_file.invoke(compiler.cljc:1642)
16:37:14  	at cljs.closure$compile_file.invokeStatic(closure.clj:653)
16:37:14  	at cljs.closure$compile_file.invoke(closure.clj:631)
16:37:14  	at cljs.closure$fn__5388.invokeStatic(closure.clj:727)
16:37:14  	at cljs.closure$fn__5388.invoke(closure.clj:721)
16:37:14  	at cljs.closure$fn__5301$G__5294__5308.invoke(closure.clj:549)
16:37:14  	at cljs.closure$compile_sources$iter__5514__5518$fn__5519.invoke(closure.clj:1088)
16:37:14  	at clojure.lang.LazySeq.sval(LazySeq.java:42)
16:37:14  	at clojure.lang.LazySeq.seq(LazySeq.java:51)
16:37:14  	at clojure.lang.Cons.next(Cons.java:39)
16:37:14  	at clojure.lang.RT.next(RT.java:709)
16:37:14  	at clojure.core$next__5371.invokeStatic(core.clj:64)
16:37:14  	at clojure.core$dorun.invokeStatic(core.clj:3142)
16:37:14  	at clojure.core$doall.invokeStatic(core.clj:3148)
16:37:14  	at clojure.core$doall.invoke(core.clj:3148)
16:37:14  	at cljs.closure$compile_sources.invokeStatic(closure.clj:1084)
16:37:14  	at cljs.closure$compile_sources.invoke(closure.clj:1073)
16:37:14  	at cljs.closure$build.invokeStatic(closure.clj:3011)
16:37:14  	at cljs.closure$build.invoke(closure.clj:2919)
16:37:14  	at cljs.build.api$build.invokeStatic(api.clj:208)
16:37:14  	at cljs.build.api$build.invoke(api.clj:189)
16:37:14  	at cljs.build.api$build.invokeStatic(api.clj:195)
16:37:14  	at cljs.build.api$build.invoke(api.clj:189)
16:37:14  	at user$eval345.invokeStatic(form-init8805134699066762225.clj:1)
16:37:14  	at user$eval345.invoke(form-init8805134699066762225.clj:1)
16:37:14  	at clojure.lang.Compiler.eval(Compiler.java:7176)
16:37:14  	at clojure.lang.Compiler.eval(Compiler.java:7166)
16:37:14  	at clojure.lang.Compiler.load(Compiler.java:7635)
16:37:14  	at clojure.lang.Compiler.loadFile(Compiler.java:7573)
16:37:14  	at clojure.main$load_script.invokeStatic(main.clj:452)
16:37:14  	at clojure.main$init_opt.invokeStatic(main.clj:454)
16:37:14  	at clojure.main$init_opt.invoke(main.clj:454)
16:37:14  	at clojure.main$initialize.invokeStatic(main.clj:485)
16:37:14  	at clojure.main$null_opt.invokeStatic(main.clj:519)
16:37:14  	at clojure.main$null_opt.invoke(main.clj:516)
16:37:14  	at clojure.main$main.invokeStatic(main.clj:598)
16:37:14  	at clojure.main$main.doInvoke(main.clj:561)
16:37:14  	at clojure.lang.RestFn.applyTo(RestFn.java:137)
16:37:14  	at clojure.lang.Var.applyTo(Var.java:705)
16:37:14  	at clojure.main.main(main.java:37)
16:37:14  Caused by: clojure.lang.ExceptionInfo: null #:clojure.error{:source nil, :line nil, :column nil, :phase :compilation}
16:37:14  	at cljs.analyzer$analyze.invokeStatic(analyzer.cljc:4128)
16:37:14  	at cljs.analyzer$analyze.invoke(analyzer.cljc:4114)
16:37:14  	at cljs.compiler$emit_source.invokeStatic(compiler.cljc:1524)
16:37:14  	at cljs.compiler$emit_source.invoke(compiler.cljc:1497)
16:37:14  	at cljs.compiler$compile_file_STAR_$fn__3870.invoke(compiler.cljc:1609)
16:37:14  	at cljs.compiler$with_core_cljs.invokeStatic(compiler.cljc:1417)
16:37:14  	at cljs.compiler$with_core_cljs.invoke(compiler.cljc:1406)
16:37:14  	at cljs.compiler$compile_file_STAR_.invokeStatic(compiler.cljc:1593)
16:37:14  	at cljs.compiler$compile_file_STAR_.invoke(compiler.cljc:1586)
16:37:14  	at cljs.compiler$compile_file$fn__3901.invoke(compiler.cljc:1691)
16:37:14  	... 41 more
16:37:14  Caused by: clojure.lang.ExceptionInfo: No such namespace: status-im.transport.message.public-chat, could not locate status_im/transport/message/public_chat.cljs, status_im/transport/message/public_chat.cljc, or JavaScript source providing "status-im.transport.message.public-chat" (Please check that namespaces with dashes use underscores in the ClojureScript file name) in file src/status_im/contact_code/core.cljs {:tag :cljs/analysis-error}
16:37:14  	at cljs.analyzer$error.invokeStatic(analyzer.cljc:751)
16:37:14  	at cljs.analyzer$error.invoke(analyzer.cljc:747)
16:37:14  	at cljs.analyzer$error.invokeStatic(analyzer.cljc:749)
16:37:14  	at cljs.analyzer$error.invoke(analyzer.cljc:747)
16:37:14  	at cljs.analyzer$analyze_deps.invokeStatic(analyzer.cljc:2567)
16:37:14  	at cljs.analyzer$analyze_deps.invoke(analyzer.cljc:2541)
16:37:14  	at cljs.analyzer$ns_side_effects.invokeStatic(analyzer.cljc:3993)
16:37:14  	at cljs.analyzer$ns_side_effects.invoke(analyzer.cljc:3988)
16:37:14  	at cljs.analyzer$analyze_STAR_$fn__2657.invoke(analyzer.cljc:4112)
16:37:14  	at clojure.lang.PersistentVector.reduce(PersistentVector.java:343)
16:37:14  	at clojure.core$reduce.invokeStatic(core.clj:6827)
16:37:14  	at clojure.core$reduce.invoke(core.clj:6810)
16:37:14  	at cljs.analyzer$analyze_STAR_.invokeStatic(analyzer.cljc:4112)
16:37:14  	at cljs.analyzer$analyze_STAR_.invoke(analyzer.cljc:4102)
16:37:14  	at cljs.analyzer$analyze.invokeStatic(analyzer.cljc:4131)
16:37:14  	... 50 more
16:37:14  Error encountered performing task 'doo' with profile(s): 'test'
16:37:14  Subprocess failed
```

cammellos commented:
@asemiankevich built successfully, thanks!

asemiankevich commented:
@cammellos i will post the findings here in this comment if u dont mind

- [ ] Profile -> about throws an error . Works fine in  nightly

![image](https://user-images.githubusercontent.com/28752642/59444753-133a1680-8e07-11e9-9611-dc26e5461a9a.png)

cammellos commented:
@asemiankevich should be fixed now, thanks!

churik commented:
Testing is in progress, so far:
#### 1) Quoted messages are shown as ordinary messages for receiver (if sent / received on release 0.13.0)
**Steps:**
- create account on release 0.13.0
- chat with account on current PR
- send message 
- reply to this message
- check  chat on release 0.13.0 and current PR

**Expected result:** quoted message is shown on both devices
**Actual result:** message is shown as ordinary for the receiver

cammellos commented:
@churik I have received a message about replies, is it an issue with this PR or develop?

churik commented:
@cammellos testing is still in progress, so far:
#### 1) "Array size is not a small enough positive integer" if you add contact after pairing
It is the same as #8355 - but in PR it is reproducible constantly with given steps.
**Steps:**
- create account on Account A
- restore this account on another device
- add any acount via public chat > tapping user photo

**Expected result:** no error
**Actual result:** 
![IMG_E0D74375127D-1](https://user-images.githubusercontent.com/4557972/59604361-44b72880-910c-11e9-9dc3-d11caba815ec.jpeg)
**Logs:** 
[status_logs.zip](https://github.com/status-im/status-react/files/3296921/status_logs.zip)

churik commented:
> @churik I have received a message about replies, is it an issue with this PR or develop?

it was with develop, sorry. I edited comment now with new issues.

cammellos commented:
@churik thanks for the feedback

2 is fixed, I had disabled it for testing and forgot to re-enable it
1 not sure what it is,I can keep looking, but might not be related to this PR, let me know if you can replicate in nightly / develop.

churik commented:
@cammellos 
rechecked issue 1 on last nightly (14-Jun-2019), it is not reproducible with the same steps.
In PR it is very visible and reproducible, on nightly it is random (as described in https://github.com/status-im/status-react/issues/8355)

New issues:
## Issue 3: Messages from the same account are shown in public chat are shown as messages sent from a different account

**Steps:**
- create an account on device A
- restore this account on device B
- join to some public chat
- send some message in public chat from device A
- reopen app on device B
 
**Expected result:**  can see messages as sender, without picture
**Actual result:** messages are shown as sent from another user
<img width="432" alt="photo_2019-06-17 16 28 51 jpeg 2019-06-17 16-29-34" src="https://user-images.githubusercontent.com/4557972/59612380-43422c00-911d-11e9-94da-a484f666ef6c.png">

## Issue 4: No gaps are shown for group chats
Use build test build #22 with default time for fetching history 60sec.
**Steps:**
- create account 
- add user B to contact, start group chat with him
- close app for 2 mins
- check group chat

**Expected result:**  can see "Fetch messages", as for 1-1 and public chats
**Actual result:**  no gaps

cammellos commented:
Thanks for the feedback @churik , I will take a look. For issue 3 is also in nightly (same repro steps), I think it might have been introduced https://github.com/status-im/status-react/pull/8336 as it touches that part of the codebase, but can't confirm, so might be worth opening a separate issue, as not sure I will have time to look into it.

cammellos commented:
@churik I have addressed 1, 3 and 4.
1 & 3 are also present on develop, so I will probably port the fixes there when I have some time, but otherwise is ready to be tested again. thanks!

cammellos commented:
@churik I have created separate PRs for the public chat messages #8431 and array range error #8432 , the fixes are included in this PR as well

churik commented:
@cammellos thanks a lot, sorry I missed somehow https://github.com/status-im/status-react/pull/8363#issuecomment-502800360.
Will recheck ASAP, and probably this PR will be merged sooner than fixes.

churik commented:
@cammellos thank you, issues 1, 3 and 4 are fixed.
## Issues

About **Issue 2: Messages in 1-1 chat are not syncing after pairing**:
now messages sent in 1-1 chats in one device are shown as messages in chat with user itself.
Issue is not reproducible in nightly.

**Steps (scenario 1):**
- as UserA: create account on device A
- as UserA: restore this account on device B
- as UserA: pair devices
- as UserA: send some message in 1-1 chat with userB from device A
- as UserA: send some message in 1-1 chat with userC from device A

**Expected result:**  can see messages from device A on device B in different chats with userB and userC accordingly
**Actual result:** messages from userB and userC are shown in 1-1 chat UserA with UserA
<img width="242" alt="Monosnap 2019-06-19 12-07-48" src="https://user-images.githubusercontent.com/4557972/59757015-24fa3e80-928b-11e9-9513-b7cd1bc44448.png">

## Questions

Please let me know If I missed smth in tests from your POV, below is the brief summary:
### Tested (IOS 11.4.1 (IPhone 7), Android 8 (LG V20)):
#### 1) Group chats
- declining invite
- chatting PR <-> PR
- chatting PR <-> Release
- deleting chat as non-admin
- make admin
- remove from chat
- clear history for chat 

#### 2) Private chats
- in PR <-> PR, PR <-> release
- messaging when not added to contacts
- PNS
- /gfycat
- /send, /request commands

#### 3) General
- search in chats
- offline inboxing
- deleting/Creating one-to-one/group/public chats
- enable / disable bootnodes
- switching networks / mailservers
- save password
- enable biometric authentication
- enabling privacy preview mode
- enabling biometric authentication
- save password
- send logs

#### 4) Pairing for (PR <-> PR, PR <-> release):
- group chats
- public chats
- contacts 
- 1-1
- photo / username 

#### 5) Retrieving history from mailserver (retrived for several public chats for 30 days)
#### 6) Creating/restoring accounts (also with password with special chars)
#### 7) Gaps in all types of chats


### What's left:
- basic desktop check
- block user feature (it is tested partially by e2e, but i think it should be tested thoroughly)
- retest fixes

cammellos commented:
thanks, sounds good for the scenarios, about one to one and pairing I
thought I had fixed the issue in the latest build, did you use that one?
otherwise i will take a look at it tonight

On Wed, Jun 19, 2019, 12:19 Tetiana Churikova <notifications@github.com>
wrote:

> @cammellos <https://github.com/cammellos> thank you, issues 1, 3 and 4
> are fixed.
> Issues
>
> About *Issue 2: Messages in 1-1 chat are not syncing after pairing*:
> now messages sent in 1-1 chats in one device are shown as messages in chat
> with user itself.
> Issue is not reproducible in nightly.
>
> *Steps (scenario 1):*
>
>    - as UserA: create account on device A
>    - as UserA: restore this account on device B
>    - as UserA: pair devices
>    - as UserA: send some message in 1-1 chat with userB from device A
>    - as UserA: send some message in 1-1 chat with userC from device A
>
> *Expected result:* can see messages from device A on device B in
> different chats with userB and userC accordingly
> *Actual result:* messages from userB and userC are shown in 1-1 chat
> UserA with UserA
> [image: Monosnap 2019-06-19 12-07-48]
> <https://user-images.githubusercontent.com/4557972/59757015-24fa3e80-928b-11e9-9513-b7cd1bc44448.png>
> Questions
>
> Please let me know If I missed smth in tests from your POV, below is the
> brief summary:
> Tested (IOS 11.4.1 (IPhone 7), Android 8 (LG V20)): 1) Group chats
>
>    - declining invite
>    - chatting PR <-> PR
>    - chatting PR <-> Release
>    - deleting chat as non-admin
>    - make admin
>    - remove from chat
>    - clear history for chat
>
> 2) Private chats
>
>    - in PR <-> PR, PR <-> release
>    - messaging when not added to contacts
>    - PNS
>    - /gfycat
>    - /send, /request commands
>
> 3) General
>
>    - search in chats
>    - offline inboxing
>    - deleting/Creating one-to-one/group/public chats
>    - enable / disable bootnodes
>    - switching networks / mailservers
>    - save password
>    - enable biometric authentication
>    - enabling privacy preview mode
>    - enabling biometric authentication
>    - save password
>    - send logs
>
> 4) Pairing for (PR <-> PR, PR <-> release):
>
>    - group chats
>    - public chats
>    - contacts
>    - 1-1
>    - photo / username
>
> 5) Retrieving history from mailserver (retrived for several public chats
> for 30 days) 6) Creating/restoring accounts (also with password with
> special chars) 7) Gaps in all types of chats What's left:
>
>    - basic desktop check
>    - block user feature (it is tested partially by e2e, but i think it
>    should be tested thoroughly)
>    - retest fixes
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/status-im/status-react/pull/8363?email_source=notifications&email_token=AAHYJMC7NCNI2WCQOTFOOIDP3IBZZA5CNFSM4HTWA5KKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODYBM5LQ#issuecomment-503500462>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAHYJMAPT7NXLBXEKWATRSTP3IBZZANCNFSM4HTWA5KA>
> .
>

annadanchenko commented:
@churik @cammellos does it make sense to test upgrade from 0.12.2 to this Pr?

cammellos commented:
yes, definitely

On Wed, Jun 19, 2019, 12:37 Anna Danchenko <notifications@github.com> wrote:

> @churik <https://github.com/churik> @cammellos
> <https://github.com/cammellos> does it make sense to test upgrade from
> 0.12.2 to this Pr?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/status-im/status-react/pull/8363?email_source=notifications&email_token=AAHYJMBXVWLYCRUGFPGGJWLP3ID5LA5CNFSM4HTWA5KKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODYBOIFY#issuecomment-503505943>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAHYJMCD4XTPG2AXOCFUFZDP3ID5LANCNFSM4HTWA5KA>
> .
>

churik commented:
@cammellos https://github.com/status-im/status-react/pull/8363#issuecomment-503500462 is based ob on last #26 build with https://github.com/status-im/status-react/commit/498f2ea4d9c303dc7cfb4f556cb0eef6d06a79b6, so it is not fixed there.

churik commented:
@cammellos and one more:
### Issue 5: Migration is failed when login after upgrade (on some accounts)
Got an error after upgrade from 0.10.0 to current PR:

![IMG_D35CD543B503-1](https://user-images.githubusercontent.com/4557972/59765474-7102ae80-929e-11e9-906e-b673d7b76ac3.jpeg)
#### Before upgrade 
Test data on Account1:
- custom userpicture
- added 5 contacts
- 1 contact is blocked
- have incoming add to contacts request that is not answered (add to contacts shown in 1x1 chat)
- 1-1 chat with a contact. With txt, emoji messages. Long messages (more than 15 lines). Links. Universal links. Send and request messages (in different currency). One of request as active (do not send eth to it). 
- Quoted messages (1x1, group, public chats)
- Dapp used, so navigation history is present ('Back' and 'FWD' webview buttons are active)
- Dapp permissions
- Transaction history has ETH and some tokens on Ropsten.
- Several of chats  are deleted (1:1, group and public chat)
- History for some chats is cleared (1:1, group and public chat)
- Unread messages (1:1, group and public chat)
- Group chat with 2 contacts with text and emoji messages (you are admin)
- Participated in group chat where you are not admin
- Messages - gifs in 1:1, public and group chats

Test data on Account2:
- restored account
- save password 
- several paired devices
- enabled preview privacy mode
#### After upgrade 
- could open app and use Account2 normally
- logged out from  Account2
- tried to login to Account1

**Logs:**  [status_logs.zip](https://github.com/status-im/status-react/files/3306104/status_logs.zip)

Please, tell me if I can help to debug it somehow.

cammellos commented:
thanks, i think it might not be due to this PR ( the error is in the
previous migration), could you try to replicate on nightly if you have time?

thanks

On Wed, Jun 19, 2019, 14:41 Tetiana Churikova <notifications@github.com>
wrote:

> @cammellos <https://github.com/cammellos> and one more:
> Issue 5: Migration is failed when login after upgrade (on some accounts)
>
> Got an error after upgrade from 0.10.0 to current PR:
>
> [image: IMG_D35CD543B503-1]
> <https://user-images.githubusercontent.com/4557972/59765474-7102ae80-929e-11e9-906e-b673d7b76ac3.jpeg>
> Before upgrade
>
> Test data on Account1:
>
>    - custom userpicture
>    - added 5 contacts
>    - 1 contact is blocked
>    - have incoming add to contacts request that is not answered (add to
>    contacts shown in 1x1 chat)
>    - 1-1 chat with a contact. With txt, emoji messages. Long messages
>    (more than 15 lines). Links. Universal links. Send and request messages (in
>    different currency). One of request as active (do not send eth to it).
>    - Quoted messages (1x1, group, public chats)
>    - Dapp used, so navigation history is present ('Back' and 'FWD'
>    webview buttons are active)
>    - Dapp permissions
>    - Transaction history has ETH and some tokens on Ropsten.
>    - Several of chats are deleted (1:1, group and public chat)
>    - History for some chats is cleared (1:1, group and public chat)
>    - Unread messages (1:1, group and public chat)
>    - Group chat with 2 contacts with text and emoji messages (you are
>    admin)
>    - Participated in group chat where you are not admin
>    - Messages - gifs in 1:1, public and group chats
>
> Test data on Account2:
>
>    - restored account
>    - save password
>    - several paired devices
>    - enabled preview privacy mode
>
> After upgrade
>
>    - could open app and use Account2 normally
>    - logged out from Account2
>    - tried to login to Account1
>
> *Logs:* status_logs.zip
> <https://github.com/status-im/status-react/files/3306104/status_logs.zip>
>
> Please, tell me if I can help to debug it somehow.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/status-im/status-react/pull/8363?email_source=notifications&email_token=AAHYJMB4BVOVWB2N6TGID7DP3ISRFA5CNFSM4HTWA5KKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODYBXPVY#issuecomment-503543767>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAHYJMBYM5HIYJ4MO644Y3LP3ISRFANCNFSM4HTWA5KA>
> .
>

churik commented:
@cammellos  yeah, it can be.
Can you point me what data can cause it just to make it a bit faster?
2 accounts? chats? blocked users? extensions?

cammellos commented:
let me have a quick look ( on mobile so might take a bit)

On Wed, Jun 19, 2019, 14:57 Tetiana Churikova <notifications@github.com>
wrote:

> @cammellos <https://github.com/cammellos> yeah, it can be.
> Can you point me what data can cause it just to make it a bit faster?
> 2 accounts? chats? blocked users? extensions?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/status-im/status-react/pull/8363?email_source=notifications&email_token=AAHYJMHJBX3ZDKQLQWSDBZDP3IUMFA5CNFSM4HTWA5KKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODYBYZQY#issuecomment-503549123>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAHYJMH2S24QXCXXMACLJFLP3IUMFANCNFSM4HTWA5KA>
> .
>

cammellos commented:
can't figure it out exactly how to replicate, but i am sure is already in
develop.
 In any case it needs fixing and the fix is simple, ( some messages don't
have a field that is being migrated, but don't know how to get one of those
messages).

i will push a fix tonight and open a separate pr as well, in case new
issues are found in this one, good catch!
I will fix it tonight

On Wed, Jun 19, 2019, 15:00 Andrea Maria Piana <andrea.maria.piana@gmail.com>
wrote:

> let me have a quick look ( on mobile so might take a bit)
>
> On Wed, Jun 19, 2019, 14:57 Tetiana Churikova <notifications@github.com>
> wrote:
>
>> @cammellos <https://github.com/cammellos> yeah, it can be.
>> Can you point me what data can cause it just to make it a bit faster?
>> 2 accounts? chats? blocked users? extensions?
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/status-im/status-react/pull/8363?email_source=notifications&email_token=AAHYJMHJBX3ZDKQLQWSDBZDP3IUMFA5CNFSM4HTWA5KKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODYBYZQY#issuecomment-503549123>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AAHYJMH2S24QXCXXMACLJFLP3IUMFANCNFSM4HTWA5KA>
>> .
>>
>

churik commented:
And I think the last one:
### Issue 6: `status_im.transport.utils.ssh .... is not a function` error when log in to status-desktop
Not reproducible on nightly build.
If upgrade to PR -> log in, see error:
<img width="568" alt="Monosnap 2019-06-19 16-37-09" src="https://user-images.githubusercontent.com/4557972/59777059-31df5800-92b4-11e9-927c-ca35d3461d28.png">
**Logs:**
[Status-debug-logs.zip](https://github.com/status-im/status-react/files/3306697/Status-debug-logs.zip)
**Platform:** Mac OSx (didn't try on others)

cammellos commented:
@churik The issues should be addressed, I could not test the migration as I don't quite know how to get an old build for Status PR, but I believe it should be fixed.

I think it has to do with delete chats/messages, so probably to replicate is you could have a bunch of 1-to-1 messages in a chat before migration, delete them (either through clear history or delete-chat) and then migrate, and the error might pop up (haven't tried to replicate though).
thanks!

churik commented:
@cammellos tried with deleted \ cleared messages in 1-1, save the password and several accounts - couldn't reproduce on nightly.
Anyway, I rechecked with the similar dataset from scratch - and couldn't reproduce the issue any more in PR, so hopefully, it is fixed.
Issue 6 with desktop is also fixed

Thank you for excellent work!
However, device paring now is working in another way. Let me know if I misunderstood  or misssed smth.

### Issue 7: Messages in group and 1-1 chats are syncing without pairing

Steps:
- create accountA on deviceA
- restore accountA on deviceB
- deviceA: start 1-1 chat, send several messages
- check deviceB
I didn't expect in this case that 1-1 chat will appear on deviceB, because I didn't pair devices.
However, 1-1 chats (and group chats as well) appeared in this case and all messages sent from deviceA are visible on deviceB.

So basically enabling/disabling device doesn't affect sync messages in 1-1 and group chats.
Also message "New device detected" appears immediately after login (before it appeared after clicking on another device "Advertise")

### Issue 8: If you add to contact via 1-1 chat after syncing, new contact doesn't appear on paired device
Probably consequence of issue 7
Steps:
- create accountA on deviceA
- restore accountA on deviceB
- deviceA: start 1-1 chat, send several messages (don't add user to contacts)
- pair deviceA and deviceB
- tap on sync all
- on deviceA go back to 1-1 chat, tap on "Add to contacts"
**Expected result:** contact is added on both devices
**Actual result:** contact is added only on deviceA

cammellos commented:
@churik both 7 and 8 seems to be new issue, I will take a look at them, thanks!

cammellos commented:
@churik I believe I have addressed the issues, although I noticed that some code in status-go is really underperformant, and I will have to fix that before this gets merged, as it's very noticeable on large accounts, so this will take some time ( I am on holiday until the end of the week), I'll let you know when is ready to be tested again.
thanks for the work so far!

cammellos commented:
@churik it should be ready to be tested again
7 is fixed 

> Also message "New device detected" appears immediately after login (before it appeared after clicking on another device "Advertise")

The message appears as soon as we detect some activity from another device, so it might happen before you click on Advertise.

I could not replicate 8, it might be due to a dropped message, currently we don't resend them.

There's an outstanding issue on upgrade that on slow devices it might take some time (1/2 minutes, depending on how many public-chat/one-to-one), but after the initial upgrade everything should be working fast (this is due to the fact that we have to generate some keys the first time around).
I can make it faster, but I would address it in a separate PR if it's ok.

Thanks for your patience

churik commented:
@cammellos about recent changes: as I understand, I should retest once more:
- upgrade
- the issues are not reproducible
- and the feature itself (from the description).
Is it right?

cammellos commented:
yes, that is correct.

The feature itself (from the logs), is not that important as of now (I have
checked many times, and I am more concerned about potential bugs), but of
course if you have time it s all good.

thanks

On Fri, Jun 21, 2019, 09:35 Tetiana Churikova <notifications@github.com>
wrote:

> @cammellos <https://github.com/cammellos> about recent changes: as I
> understand, I should retest once more:
>
>    - upgrade
>    - the issues are not reproducible
>    - and the feature itself (from the description).
>    Is it right?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/status-im/status-react/pull/8363?email_source=notifications&email_token=AAHYJMCPXUEBGKZ3RJRCNLDP3SAFNA5CNFSM4HTWA5KKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODYHWGQY#issuecomment-504324931>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAHYJMDNBXXUPBND2JKLZBDP3SAFNANCNFSM4HTWA5KA>
> .
>

churik commented:
@cammellos 
Issue 7 is fixed.
Issue 8 is still reproducible, more detailed steps are below (sorry if they weren't clear):
### Issue 8: If you add to contact via 1-1 chat after syncing, new contact doesn't appear on paired device

**Steps:**
- create accountA on deviceA
- restore accountA on deviceB
- deviceA: start 1-1 chat with UserC, send several messages to UserC (don't add UserC to contacts)
So "Add to contacts" should be visible in 1-1 chat.
- deviceB: advertise deviceB to deviceA
- deviceA: enable deviceB in "Devices"
- deviceA: tap on sync all
- deviceA: send message to  1-1 chat with UserC. It should appear on deviceB.
- deviceA :  tap on "Add to contacts" in 1-1 chat with UserC

**Expected result:** contact is added on both devices
**Actual result:** contact is added only on deviceA.

 "Sync all" on deviceA doesn't help.
All fine if I add new contact on deviceA from public chat - it appears on deviceB.

**Logs from deviceA:**  [Status-debug-logsA.zip](https://github.com/status-im/status-react/files/3313756/Status-debug-logsA.zip)

**Logs from deviceB:** [status_logsB.zip](https://github.com/status-im/status-react/files/3313758/status_logsB.zip)

cammellos commented:
thanks, i will take another look at it tonight, so probably will be ready
on Monday as i am off today. have a good weekend!

On Fri, Jun 21, 2019, 11:41 Tetiana Churikova <notifications@github.com>
wrote:

> @cammellos <https://github.com/cammellos>
> Issue 7 is fixed.
> Issue 8 is still reproducible, more detailed steps are below (sorry if
> they weren't clear):
> Issue 8: If you add to contact via 1-1 chat after syncing, new contact
> doesn't appear on paired device
>
> *Steps:*
>
>    - create accountA on deviceA
>    - restore accountA on deviceB
>    - deviceA: start 1-1 chat with UserC, send several messages to UserC
>    (don't add UserC to contacts)
>    So "Add to contacts" should be visible in 1-1 chat.
>    - deviceB: advertise deviceB to deviceA
>    - deviceA: enable deviceB in "Devices"
>    - deviceA: tap on sync all
>    - deviceA: send message to 1-1 chat with UserC. It should appear on
>    deviceB.
>    - deviceA : tap on "Add to contacts" in 1-1 chat with UserC
>
> *Expected result:* contact is added on both devices
> *Actual result:* contact is added only on deviceA.
>
> "Sync all" on deviceA doesn't help.
> All fine if I add new contact on deviceA from public chat - it appears on
> deviceB.
>
> *Logs from deviceA:* Status-debug-logsA.zip
> <https://github.com/status-im/status-react/files/3313756/Status-debug-logsA.zip>
>
> *Logs from deviceB:* status_logsB.zip
> <https://github.com/status-im/status-react/files/3313758/status_logsB.zip>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/status-im/status-react/pull/8363?email_source=notifications&email_token=AAHYJMDZCONBOFZ7EV3EJ2TP3SO23A5CNFSM4HTWA5KKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODYH7TCI#issuecomment-504363401>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAHYJMGGEZOHNCPZAOOK3V3P3SO23ANCNFSM4HTWA5KA>
> .
>

cammellos commented:
@churik thanks for the repro steps, managed to replicate.

I believe the issue is also present in develop, and likely was introduced either with the introduction of system-tags or tribute to talk, as essentially system tags have changed the way we represent a contact, while ttt made some changes on when we store a record (opening a chat will add a "contact" to a list internally, which was then synced when pressing `sync-all`), and pairing code was not updated accordingly.
In any case it should be fixed in the latest build, good catch!

There might still be some issues lurking around contacts & syncing, but probably best to fix separately.

churik commented:
Retested on previous builds: 
- fresh install: cases from https://github.com/status-im/status-react/pull/8363#issuecomment-503500462
-  upgrade: IOS (0.13.0 -> PR build) with data https://github.com/status-im/status-react/pull/8363#issuecomment-503543767

Retested on last build (https://github.com/status-im/status-react/commit/252ab1ac52045f385edadcf886cb36a115110f2e):
- triggered a full bunch of e2e https://ethstatus.testrail.net/index.php?/runs/view/3633 - no difference with nightly 24/07/2019.
- pairing (IOS, Android)
- compatibility with 0.13.0
- basic check of desktop app (Mac OS)
- upgrade to release build in PR brunch with data https://github.com/status-im/status-react/pull/8363#issuecomment-503543767 on  Android 7.0

Looks good for me.
Thanks @cammellos for your patience!

cammellos commented:
awesome, thanks a lot for your help and patience @churik !
